File: js/layoutEngine.js
// Layout Engine
const CANVAS_WIDTH = 1240;
const CANVAS_HEIGHT = 1754;

/**
 * Global canvas specifications for A4 manga pages.
 * @type {Object}
 * @property {number} width - The width of the canvas in pixels.
 * @property {number} height - The height of the canvas in pixels.
 */
export const CanvasSpecs = {
    width: CANVAS_WIDTH,
    height: CANVAS_HEIGHT
};

/**
 * Calculates the panel coordinates and dimensions based on the requested layout type.
 *
 * @param {string} layoutType - The type of layout to generate. Options: "splash", "grid", "cinematic".
 * @returns {Array<{id: number, x: number, y: number, width: number, height: number}>}
 *          - An array of panel objects containing position (x, y) and dimensions (width, height).
 *          Returns a "splash" layout by default if the layout type is unknown.
 */
export function calculateLayout(layoutType) {
    const type = layoutType.toLowerCase();
    const panels = [];
    const margin = 50;
    const gap = 20;

    const availableWidth = CANVAS_WIDTH - (2 * margin);
    const availableHeight = CANVAS_HEIGHT - (2 * margin);

    if (type === "splash") {
        panels.push({
            id: 1,
            x: margin,
            y: margin,
            width: availableWidth,
            height: availableHeight
        });
    } else if (type === "grid") {
        const pWidth = Math.floor((availableWidth - gap) / 2);
        const pHeight = Math.floor((availableHeight - gap) / 2);

        // Panel 1 (Top Left)
        panels.push({ id: 1, x: margin, y: margin, width: pWidth, height: pHeight });
        // Panel 2 (Top Right)
        panels.push({ id: 2, x: margin + pWidth + gap, y: margin, width: pWidth, height: pHeight });
        // Panel 3 (Bottom Left)
        panels.push({ id: 3, x: margin, y: margin + pHeight + gap, width: pWidth, height: pHeight });
        // Panel 4 (Bottom Right)
        panels.push({ id: 4, x: margin + pWidth + gap, y: margin + pHeight + gap, width: pWidth, height: pHeight });
    } else if (type === "cinematic") {
        const pWidth = availableWidth;
        const pHeight = Math.floor((availableHeight - (2 * gap)) / 3);

        // Panel 1 (Top)
        panels.push({ id: 1, x: margin, y: margin, width: pWidth, height: pHeight });
        // Panel 2 (Middle)
        panels.push({ id: 2, x: margin, y: margin + pHeight + gap, width: pWidth, height: pHeight });
        // Panel 3 (Bottom)
        panels.push({ id: 3, x: margin, y: margin + 2 * (pHeight + gap), width: pWidth, height: pHeight });
    } else {
        // Fallback
        console.warn(`Unknown layout '${type}', defaulting to Splash.`);
        return calculateLayout("splash");
    }

    return panels;
}

File: manga_maker/src/layout_engine.py
from typing import Dict, List, Tuple

# Canvas Specs
CANVAS_WIDTH = 1240
CANVAS_HEIGHT = 1754
BORDER_WIDTH = 5

def calculate_layout(layout_type: str) -> List[Dict[str, int]]:
    """
    Translates layout type into panel coordinates.
    Returns a list of dictionaries, each containing x, y, width, height for a panel.
    """
    layout_type = layout_type.lower()
    panels = []

    # Define margins/padding
    margin = 50
    gap = 20

    available_width = CANVAS_WIDTH - (2 * margin)
    available_height = CANVAS_HEIGHT - (2 * margin)

    if layout_type == "splash":
        # One massive panel
        panels.append({
            "x": margin,
            "y": margin,
            "width": available_width,
            "height": available_height,
            "id": 1
        })

    elif layout_type == "grid":
        # 2x2 Grid
        # Width of one panel
        p_width = (available_width - gap) // 2
        p_height = (available_height - gap) // 2

        # Panel 1 (Top Left)
        panels.append({"x": margin, "y": margin, "width": p_width, "height": p_height, "id": 1})
        # Panel 2 (Top Right)
        panels.append({"x": margin + p_width + gap, "y": margin, "width": p_width, "height": p_height, "id": 2})
        # Panel 3 (Bottom Left)
        panels.append({"x": margin, "y": margin + p_height + gap, "width": p_width, "height": p_height, "id": 3})
        # Panel 4 (Bottom Right)
        panels.append({"x": margin + p_width + gap, "y": margin + p_height + gap, "width": p_width, "height": p_height, "id": 4})

    elif layout_type == "cinematic":
        # Three wide, stacked panels
        p_width = available_width
        p_height = (available_height - (2 * gap)) // 3

        # Panel 1 (Top)
        panels.append({"x": margin, "y": margin, "width": p_width, "height": p_height, "id": 1})
        # Panel 2 (Middle)
        panels.append({"x": margin, "y": margin + p_height + gap, "width": p_width, "height": p_height, "id": 2})
        # Panel 3 (Bottom)
        panels.append({"x": margin, "y": margin + 2 * (p_height + gap), "width": p_width, "height": p_height, "id": 3})

    else:
        # Default to Splash if unknown
        print(f"Unknown layout '{layout_type}', defaulting to Splash.")
        return calculate_layout("splash")

    return panels
